### PR TITLE
Add pager

### PR DESCRIPTION
**Pull request by chooglen**
Add pager support via the following config variables:

- ui.pager: the program to use as the pager. Defaults to $PAGER.
- ui.paginate: whether to paginate output. Supported values are 'auto',
  'always' and 'never'. Defaults to 'auto'.

Support for the `--pager` (or should it be called `--paginate`?) flag is WIP.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)

### CODE REVIEW

File: src/ui.rs [hunk 1]
```diff
@@ -170,6 +233,30 @@ impl Ui {
...
+        match &self.output_pair {
+            UiOutput::Paged { buffer } => {
+                if buffer.lock().unwrap().is_empty() {
+                    Ok(())
+                } else {
+                    let pager_cmd = pager_setting(self.settings());
+                    let mut pager = Command::new(pager_cmd)
+                        .stdin(Stdio::piped())
+                        .spawn()
+                        .expect("Failed to spawn pager");
```
Comment (martinvonz) [inline src/ui.rs:246]: We need to start the pager earlier. If we start it only at the end, commands that can produce a lot of ouput (such as `jj log`) will need to finish producing *all* their output before the user sees any of it. In addition to high latency, it's also wasteful to produce all the output if the user just wants a few pages. FYI, the way this currently works if you do `jj log | less` (or use git's or hg's log paginated commands) is that the commands just keep producing output but the pager typically makes the writes block after a certain amount, which makes the main process also wait and not produce too much output.

**Comment by martinvonz** [comment]
> * e.g. maybe we really like `struct Ui`

Yes, I think the goal should be to route `clap`'s output through the `Ui` object. We have previously talked about also avoiding having `clap` exit the process (by calling `try_get_matches_from()` instead of `get_matches_from()`).

> * How we could change the default paging behavior per-command

Yes, that's what Mercurial does. For many commands (e.g. `jj squash`), it doesn't make sense to spawn a pager because the output is just a few lines at most.

> Writes first go to a temporary buffer and the pager is spawned only if
the buffer is nonempty. 

Mercurial lets the commands decide when to enable the pager (if it's configured to be used). That should remove the need for a temporary buffer.

File: src/ui.rs (hunk 1 above)
Comment (chooglen) [reply src/ui.rs:246]: Makes sense. Then instead of writing to temporary buffer, we could spawn the pager the first time we try to write to it.

An alternative is to always spawn the pager, but as I noted in the commit message, spawning the pager but not writing to it has unintended side effects.

Comment (martinvonz) [reply src/ui.rs:246]: > An alternative is to always spawn the pager, but as I noted in the commit message, spawning the pager but not writing to it has unintended side effects.

What was that scenario again? Can we decide that that's a broken setup of less? I think hg seems to have done that, because I'm pretty sure they spawn the pager even if they don't end up writing to it.

Comment (chooglen) [reply src/ui.rs:246]: > What was that scenario again?
That scenario is:

- Spawn `less`
- Write to stdout, bypassing `less`

>  Can we decide that that's a broken setup of less?
I don't think this is a broken `less` setup; this is the default setup on my machine. `LESS=X` does make the problem go away by disabling termcap init/deinit, but AIUI setting this is more of a matter of personal taste than a good default, e.g. I've seen mention that it disables features like mouse scroll support.

> I think hg seems to have done that, because I'm pretty sure they spawn the pager even if they don't end up writing to it.
Perhaps this works well for hg because once the pager is spawned, all future output is guaranteed to go to the pager because stdout/stderr get dup()-ed to the pager's stdin.

Comment (martinvonz) [reply src/ui.rs:246]: > * Write to stdout, bypassing `less`

It seems that we just shouldn't do that. Do you mean the output that `clap` prints for us? If we don't start the pager until the command line has been successfully parsed, that shouldn't happen, I think. Or is there another case where we would do that?

Comment (chooglen) [reply src/ui.rs:246]: >> Write to stdout, bypassing less

> It seems that we just shouldn't do that.

If we don't ever intend to write to stdout, then isn't dup() a cleaner approach?
Or perhaps the point is that this isn't actually pager-specific; perhaps at some point we want to be able to write to arbitrary destinations in ways that might not map well to dup()-ed file descriptors.

> Do you mean the output that clap prints for us? ... Or is there another case where we would do that?

I've only observed clap doing this, and it's probably the only case where that would happen.

> If we don't start the pager until the command line has been successfully parsed, that shouldn't happen, I think.

Hm, let me try that out then.

**Comment by Ralith** [comment]
> Whether we really need DynWriteLock or if there are simpler alternatives here

We should stream output directly to the pager rather than buffering it up in full first. A file descriptor doesn't require locking to write to, so in that case `DynWriteLock` will become unnecessary.

File: src/ui.rs (hunk 1 above)
Comment (Ralith) [reply src/ui.rs:246]: I think it makes a lot of sense to spawn the pager on-demand, only for specific parts of the output of specific subcommands. For example, a long-running operation which produces a lot of output at the end might want to display a progress bar directly to the terminal before switching to a pager later on.

**Comment by yuja** [comment]
> A file descriptor (or [`ChildStdin`](https://doc.rust-lang.org/std/process/struct.ChildStdin.html)) doesn't require locking to write to, so in that case `DynWriteLock` will become unnecessary.

Yep, I lied in the chat, sorry. If we need line-buffered output (like stdout), I think `&ChildStdin` can be wrapped per `stdout_formatter()` call, instead of locking shared buffered writer.

File: src/ui.rs [hunk 2]
```diff
@@ -73,6 +75,50 @@ fn use_color(choice: ColorChoice) -> bool {
...
+        .config()
+        .get_string("ui.paginate")
+        .ok()
+        .and_then(|s| s.parse().ok())
+        .unwrap_or_default()
+}
+
+fn pager_setting(settings: &UserSettings) -> String {
+    settings.config().get_string("ui.pager").unwrap_or_default()
+}
```
[the remark this answers is not in the document]
Comment (chooglen) [reply src/ui.rs]: Hm, if we move this + the color choice stuff to `UserSettings`, then instead of doing `color_setting(settings)`, we'd have `settings.color`. That looks like it makes more sense, yeah.

[the remark this answers is not in the document]
Comment (chooglen) [reply src/ui.rs]: Yeah this looks better IMO

**Comment by chooglen** [comment]
Ok this is ready for review again :)

File: src/ui.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinvonz) [reply src/ui.rs]: We need to figure out at some point where things like this should go. It doesn't seem like it belongs in the lib crate (where `UserSettings` lives) since it's specific to the CLI. There were already a few similar cases, so I don't mean that we shouldn't add this config there; I'm just saying that we should think more about it at some point.

[the remark this answers is not in the document]
Comment (martinvonz) [reply src/ui.rs]: Hmm, but this PR moves quite a lot of code that's obviously related to the CLI into the lib crate. The previous offenders just named a config variable and maybe a default value. It seems unfortunate to add those CLI-specific types to the lib crate if we know we're going to move them out later. What do you think?

[the remark this answers is not in the document]
Comment (chooglen) [reply src/ui.rs]: I also considered that.  I moved them into the `UserSettings` because `use_progress_indicator` is there, and that seems obviously CLI-based too. As long as all of these settings are in the same place, I don't feel _too_ strongly about it. Maybe I have a slight preference for `ui.rs`; it seems a bit more accurate and preserves the status quo.

File: src/ui.rs [hunk 3]
```diff
@@ -88,13 +112,21 @@ impl Ui {
     pub fn stdout_formatter<'a>(&'a self) -> Box<dyn Formatter + 'a> {
         match &self.output {
             UiOutput::Terminal { stdout, .. } => self.new_formatter(stdout.lock()),
+            UiOutput::Paged { .. } => {
```
Comment (martinvonz) [inline src/ui.rs:115]: We already have `self.output.child` available here (just replace `..` by `child`), and we can get the compiler to check that the panic in `pager_stdin()` is impossible if we instead get the stdin from that.

Actually, maybe we should even store the `ChildStdout` and `ChildStderr` in separate fields of `UiOutput::Paged`? Creating it something like this:

```
...
let child_stdout = child.stdout.take().unwrap();
let child_stderr = child.stderr.take().unwrap();
UiOutput::Paged {
  child,
  child_stdout,
  child_stderr,
}
```

Would that work? Then we also avoid unwrapping them on every call.

File: src/ui.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (martinvonz) [reply src/ui.rs]: Right, there are already a few options that are obviously CLI-related. I also considered moving them to `ui.rs` (onto `Ui`, probably). I think that would be better.

File: src/ui.rs [hunk 4]
```diff
@@ -151,6 +186,18 @@ impl Ui {
     pub fn flush(&mut self) -> io::Result<()> {
         match &mut self.output {
             UiOutput::Terminal { stdout, .. } => stdout.flush(),
+            // Unimplemented
```
Comment (martinvonz) [inline src/ui.rs:189]: Do we need to do anything here? Or is this obsolete?

Comment (chooglen) [reply src/ui.rs:189]: AFAICT we only use this for the progress display, which won't apply to the pager. I don't know if there are plans to use `flush()` for output destined for the pager, but I suspect not.

Comment (chooglen) [reply src/ui.rs:189]: Hm, I could just call `.flush()` on the stdin, that probably does what we want.

File: src/ui.rs (hunk 3 above)
Comment (martinvonz) [reply src/ui.rs:115]: Eh, I mean only `child_stdin`, but you probably understood that :)

Comment (chooglen) [reply src/ui.rs:115]: I'm not sure I understand the point of storing the child's `stdout` and `stderr`, since we don't interact with them.
Storing the child's `stdin` makes a lot of sense though. Let me see if the borrow checker will let me, I vaguely remember that being an obstacle before.

Comment (chooglen) [reply src/ui.rs:115]: Ah, we've discussed this before - the problem is that the `UiOutput::Paged` would hold both the `child` and `child.stdin`, which isn't allowed by ownership rules (I think?). The solution we came up with was to dynamically get a reference to the stdin (aka what we do here).

File: src/ui.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (chooglen) [reply src/ui.rs]: Okay. To avoid a lot of churn, I'll move them to `ui.rs`, but I won't add the fields to `Ui` (since there's a good chance we'll just end up moving them again).

File: src/ui.rs (hunk 3 above)
Comment (martinvonz) [reply src/ui.rs:115]: `child.stdin.take().unwrap()` takes over the ownership from `child`, so I think that should work.

Comment (chooglen) [reply src/ui.rs:115]: Oh neat, https://doc.rust-lang.org/std/process/struct.Child.html#structfield.stdin uses the same trick.
This doesn't work for the formatter though, which stores a reference to the stdin, but I can inline a shortened version of this function there.

File: src/ui.rs [hunk 5]
```diff
@@ -137,7 +144,7 @@ impl Ui {
     /// Whether continuous feedback should be displayed for long-running
     /// operations
     pub fn use_progress_indicator(&self) -> bool {
-        self.settings().use_progress_indicator() && atty::is(Stream::Stdout)
+        progress_indicator_setting(self.settings()) && atty::is(Stream::Stdout)
```
Comment (martinvonz) [inline src/ui.rs:221]: Nit: Instead of passing in `self.settings()`, can the function be a member function on `Ui` (i.e. have a `&self` that's a `Ui`)?

File: src/main.rs [hunk 6]
```diff
@@ -19,7 +19,9 @@ use jujutsu::ui::Ui;
 fn run(ui: &mut Ui) -> Result<(), CommandError> {
     let app = default_app();
     let (command_helper, matches) = parse_args(ui, app, std::env::args_os())?;
-    run_command(ui, &command_helper, &matches)
+    let result = run_command(ui, &command_helper, &matches);
+    ui.finalize_writes()?;
```
Comment (martinvonz) [inline src/main.rs:23]: nit: The example applications have the call in main() instead. I suppose the advantage of having it in `run()` is that we also handle errors from `finalize_writes()`. So do the same in the example applications?

File: src/ui.rs [hunk 7]
```diff
@@ -194,6 +271,17 @@ impl Ui {
...
             UiOutput::Terminal { stdout, .. } => stdout.flush(),
+            UiOutput::Paged { child } => child.stdin.take().unwrap().flush(),
+        }
+    }
+
+    pub fn finalize_writes(&mut self) -> io::Result<()> {
+        match &mut self.output {
+            UiOutput::Paged { child } => {
+                child.wait()?;
+                Ok(())
```
Comment (martinvonz) [inline src/ui.rs:282]: Do you think we should restore the output to use the regular stdout and stderr here? I ask because I saw that `main.rs` and the example applications call this just before handling errors. Maybe we would ideally print the errors to stderr instead of printing them to the pager? Especially if the error comes from starting the pager.

File: src/ui.rs [hunk 8]
```diff
@@ -204,4 +292,17 @@ impl Ui {
...
+}
+
+impl UiOutput {
+    fn new_paged(settings: &UserSettings) -> UiOutput {
+        let pager_cmd = pager_setting(settings);
+        UiOutput::Paged {
+            child: Command::new(pager_cmd)
+                .stdin(Stdio::piped())
+                .spawn()
+                .expect("Failed to spawn pager"),
```
Comment (martinvonz) [inline src/ui.rs:305]: Would be better to return this as an error (`CommandError::UserError`), right?

File: src/ui.rs [hunk 9]
```diff
@@ -151,19 +225,22 @@ impl Ui {
         let data = text.as_bytes();
         match &mut self.output {
             UiOutput::Terminal { stdout, .. } => stdout.write_all(data),
+            UiOutput::Paged { child } => child.stdin.take().unwrap().write_all(data),
```
Comment (martinvonz) [inline src/ui.rs:228]: `child.stdin.take()` will remove the value from `child`, so this can only be called once, I think. I meant that you can store both `Child` and `ChildStdin` as separate fields in `UiOutput::Paged` if that works.

Comment (chooglen) [reply src/ui.rs:228]: > I meant that you can store both Child and ChildStdin as separate fields in UiOutput::Paged if that works

If it possible, it's quite beyond my current Rust skills 😛

Comment (martinvonz) [reply src/ui.rs:228]: I pushed https://github.com/martinvonz/jj/commit/55dc237309ad6d634428e9b89e8deb7723241a3d to show what I meant :)

File: src/main.rs (hunk 6 above)
Comment (chooglen) [reply src/main.rs:23]: Hm, actually, this is probably better in `main()`, because we have to make sure that `finalize_writes()` is always called, even if `create_ui()` or `parse_args()` fails.

File: src/ui.rs (hunk 5 above)
Comment (chooglen) [reply src/ui.rs:221]: We could, though that's inconsistent with `color_setting()` and `pagination_setting()`. Those read `UserSettings` because they're called when initializing a `Ui` (though the progess indicator isn't called that early).

File: src/ui.rs (hunk 7 above)
Comment (chooglen) [reply src/ui.rs:282]: Yeah that's a good point. Let me try that.

File: src/ui.rs (hunk 5 above)
Comment (martinvonz) [reply src/ui.rs:221]: Right, I think that's the only reason they're not methods on the `Ui` object; I think they would have been methods on it otherwise.

File: src/ui.rs (hunk 9 above)
Comment (chooglen) [reply src/ui.rs:228]: Why does it look so easy when you do it??

File: src/ui.rs (hunk 5 above)
Comment (chooglen) [reply src/ui.rs:221]: Hm, ok. Then I think this is a good enough reason to just store the pagination setting in the `Ui` struct like you initially suggested.

File: src/ui.rs (hunk 7 above)
Comment (yuja) [reply src/ui.rs:282]: You can use `std::mem::replace()` to restore stdout/stderr and take ownership of `Paged { ... }`.
That's probably needed if you made `UiOutput::Paged` hold `{ child, stdin }` pair.

https://doc.rust-lang.org/std/mem/fn.replace.html

```rust
match mem::replace(&mut self.output, UiOutput::Terminal { ... }) {
    UiOutput::Paged { ... } => {
        drop(stdin); // close()
        child.wait()?;
        ...
    }
}
```

Comment (chooglen) [reply src/ui.rs:282]: Between this and the other comment on printing an error when we fail to spawn the pager, I think it might be better to always hold a handle to stdout/stderr instead of swapping out `Ui.output` when the pager is requested.

File: src/ui.rs (hunk 8 above)
Comment (chooglen) [reply src/ui.rs:305]: Yes. Doing that makes it clear that this can cause creation of the `Ui` to fail, which is quite troubling to me because that means that we lose our only way of writing to stdout/stderr without breaking our own abstractions. Let me try to find a way that doesn't make `Ui` creation fail.

Comment (martinvonz) [reply src/ui.rs:305]: Seems like falling back to unpaged output if the user chose a bad (e.g. non-existent) pager is a good choice

Comment (martinvonz) [reply src/ui.rs:305]: To clarify, I suggest that if the call to `UiOutput::new_paged(&settings)` in `Ui::for_terminal()` fails, then we use the unpaged `UiOutput`.

Comment (yuja) [reply src/ui.rs:305]: Alternatively, maybe we can always create a plain ui and apply user settings later.
`cli_util::create_ui()` has a similar workaround.

```rust
let ui = Ui::new();
// process --color, --pager, --config, etc.
ui.reset(new_settings); // will spawn pager here
```

Anyway, I like the idea of falling back to unpaged output on error.

File: lib/src/settings.rs [hunk 10]
```diff
@@ -108,12 +108,6 @@ impl UserSettings {
             .unwrap_or(false)
     }
 
-    pub fn use_progress_indicator(&self) -> bool {
-        self.config
-            .get_bool("ui.progress-indicator")
-            .unwrap_or(true)
-    }
-
```
Comment (Ralith) [inline lib/src/settings.rs:116]: Why is this moved out, and what does it have to do with the pager?

Comment (martinvonz) [reply lib/src/settings.rs:116]: See this PR's second commit. Since this is specific to the CLI, it shouldn't be in the lib crate.

File: src/ui.rs (hunk 9 above)
Comment (chooglen) [reply src/ui.rs:228]: This sounds absurd, but when I make this change, `less` no longer recognizes the end of input when I scroll to the end. I have no idea why this might be. Let me ask for advice.

File: src/ui.rs [hunk 11]
```diff
@@ -185,8 +270,19 @@ impl Ui {
...
     pub fn flush(&mut self) -> io::Result<()> {
-        match &mut self.output_pair {
-            UiOutputPair::Terminal { stdout, .. } => stdout.flush(),
+        match &mut self.output {
+            UiOutput::Terminal { stdout, .. } => stdout.flush(),
+            UiOutput::Paged { child_stdin, .. } => child_stdin.flush(),
+        }
+    }
+
+    pub fn finalize_writes(&mut self) -> io::Result<()> {
```
Comment (Ralith) [inline src/ui.rs:279]: If this function is intended to close the `ChildStdin`, it needs to take some action to drop it, e.g. by replacing `self.output` with `UiOutput::Terminal` again, as suggested in the adjacent thread.

File: src/ui.rs (hunk 9 above)
Comment (Ralith) [reply src/ui.rs]: This is because you're no longer dropping `ChildStdin` at any point before `less` exits. See further discussion below.

File: src/ui.rs (hunk 11 above)
Comment (chooglen) [reply src/ui.rs:279]: Thanks. I think this is identical to @yuja's suggestion to drop the `ChildStdin`, i.e.

```
    pub fn finalize_writes(&mut self) -> io::Result<()> {
        match &mut self.output {
            UiOutput::Paged { child, child_stdin } => {
                # add this line
                drop(child_stdin);
                child.wait()?;
            }
            _ => (),
        };
        Ok(())
    }
```

Which _should_ fix the issue (since dropping the `ChildStdin` should `close()` the fd), but it doesn't :/

Comment (Ralith) [reply src/ui.rs:279]: It doesn't fix the issue because it doesn't drop the `ChildStdin`, only a reference to it. Because you're matching on `&mut self.output`, any pattern variables are implicitly bound to mutable references to the places they match, rather than being bound to their values. This feature is referred to as "match ergonomics" and I recommend avoiding it entirely due to this type of confusion. Always match on values rather than references, and when you want a reference to a place, request it explicitly with `ref [mut]` qualifiers on the pattern variable.

Comment (yuja) [reply src/ui.rs:279]: Yeah, I'm not against "match ergonomics", but this one is tricky. Fortunately, clippy can detect noop drop.
```
error: calls to `std::mem::drop` with a reference instead of an owned value. Dropping a reference does nothing
   --> src/ui.rs:282:17
    |
282 |                 drop(child_stdin);
    |                 ^^^^^^^^^^^^^^^^^
    |
    = note: `#[deny(clippy::drop_ref)]` on by default
note: argument has type `&mut std::process::ChildStdin`
   --> src/ui.rs:282:22
    |
282 |                 drop(child_stdin);
    |                      ^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#drop_ref
```

Comment (chooglen) [reply src/ui.rs]: Perfect. Thanks so much for the help everyone!

File: src/ui.rs [hunk 12]
```diff
@@ -240,15 +335,49 @@ impl Ui {
     pub fn output_guard(&self, text: String) -> OutputGuard {
         OutputGuard {
             text,
-            output: match self.output_pair {
-                UiOutputPair::Terminal { .. } => io::stdout(),
+            output: match self.output {
+                UiOutput::Terminal { .. } => io::stdout(),
+                UiOutput::Paged { .. } => io::stdout(),
```
Comment (chooglen) [inline src/ui.rs:340]: @Ralith I wasn't absolutely sure what to do here, but since the purpose is to write when it is dropped by the signal handler, it made sense to me to write to `io::stdout()` if we should write at all.

File: src/ui.rs [hunk 13]
```diff
@@ -116,6 +167,20 @@ impl Ui {
...
 
+    /// Switches the output to use the pager, if allowed.
+    pub fn request_pager(&mut self) {
+        if !matches!(self.paginate, PaginationChoice::Auto) {
+            return;
+        }
+        if matches!(self.output, UiOutput::Paged { .. }) {
+            return;
+        }
+        if !atty::is(Stream::Stdout) {
```
Comment (martinvonz) [inline src/ui.rs:178]: It looks like it should never make a difference, but what do you think about checking for the Stdout in `self.output`? Combined with the case above, that might look like this:

```
match &self.output {
  UiOutput::Paged { .. } => {
    return;
   }
  UiOutput::Terminal { stdout, ..} => {
    if !atty::is(stdout) {
      return;
    }
  }
}
self.output = UiOutput::new_paged_else_terminal(&self.settings);
```

Actually, maybe we should invert that to be
```
match &self.output {
  UiOutput::Paged = {}
  UiOutput::Terminal { stdout, ..} => {
    if atty::is(stdout) {
      self.output = UiOutput::new_paged_else_terminal(&self.settings);
    }
  }
}
```
You decide.

File: src/ui.rs [hunk 14]
```diff
@@ -116,6 +167,20 @@ impl Ui {
         }
     }
 
+    /// Switches the output to use the pager, if allowed.
+    pub fn request_pager(&mut self) {
+        if !matches!(self.paginate, PaginationChoice::Auto) {
```
Comment (martinvonz) [inline src/ui.rs:172]: You should be able to use `!=` here instead of `!matches!()`.

Also, I think returning if `self.paginate == PaginationChoice::Never` would be more obvious. When I saw this code, I was confused why the `Always` case wasn't handled here.

File: src/ui.rs (hunk 12 above)
Comment (Ralith) [reply src/ui.rs:340]: I think probably we don't need to do anything at all if the pager currently owns the terminal, since it'll have its own state handling, so this is fine, though we could also come up with a null output by tweaking the `OutputGuard` definition a little.

**Review by martinvonz — COMMENTED** [review]
This definitely deserves a mention in the changelog! Probably also worth mentioning in docs/config.md.

File: src/ui.rs [hunk 15]
```diff
@@ -202,8 +279,26 @@ impl Ui {
...
+        } = mem::replace(&mut self.output, UiOutput::new_terminal())
+        {
+            drop(child_stdin);
+            child
+                .wait()
+                // It's possible (though unlikely) that this write fails, but
+                // this function gets called so late that there's not much we
+                // can do about it.
+                .map_err(|e| self.write_error(&format!("Failed to wait on pager '{}'", e)))
+                .ok();
```
Comment (Ralith) [inline src/ui.rs:301]: This could be written more cleanly as `if let Err(e) = child.wait() { ... }`. Prefer to only use `map`-like functions when transforming contained data into a new form that will be processed further downstream.

File: src/ui.rs [hunk 16]
```diff
@@ -202,8 +279,26 @@ impl Ui {
...
+            child_stdin,
+        } = mem::replace(&mut self.output, UiOutput::new_terminal())
+        {
+            drop(child_stdin);
+            child
+                .wait()
+                // It's possible (though unlikely) that this write fails, but
+                // this function gets called so late that there's not much we
+                // can do about it.
+                .map_err(|e| self.write_error(&format!("Failed to wait on pager '{}'", e)))
```
Comment (Ralith) [inline src/ui.rs:300]: UI nit:
```suggestion
                .map_err(|e| self.write_error(&format!("Failed to wait on pager: {}", e)))
```

File: src/ui.rs (hunk 14 above)
Comment (chooglen) [reply src/ui.rs:172]: Separate from the fact that `Pagination::Always` _should_ imply that the pager is already active, I vaguely remember my mental model being that `request_pager` only makes sense with `PaginationChoice::Auto`, but what you said makes more sense to me right now.

File: src/ui.rs (hunk 13 above)
Comment (chooglen) [reply src/ui.rs:178]: I think `atty::is(Stream::Stdout)` is a bit clearer because it's closer to "Is the stdout connected to a tty?" as opposed to `atty::is(stdout)`, which technically is always the `stdout`, but with an extra bit of indirection through `UiOutput::Terminal`.

Also, I think `atty::is` accepts an internal `Stream` enum with variants corresponding to the standard fd-s, not the structs from `std::io`

File: src/ui.rs (hunk 12 above)
Comment (chooglen) [reply src/ui.rs:340]: Okay, I'll leave a little TODO comment then

File: src/ui.rs (hunk 14 above)
Comment (Ralith) [reply src/ui.rs:172]: This confused me a little at first but I worked out the intention you described quickly. We probably don't want to spawn a new pager if one already exists. Maybe just comment?

File: src/ui.rs [hunk 17]
```diff
@@ -84,19 +94,67 @@ fn use_color(choice: ColorChoice) -> bool {
...
+
 impl Ui {
     pub fn for_terminal(settings: UserSettings) -> Ui {
         let cwd = std::env::current_dir().unwrap();
         let color = use_color(color_setting(&settings));
+        let paginate = pagination_setting(&settings);
+        let progress_indicator = progress_indicator_setting(&settings);
         let formatter_factory = FormatterFactory::prepare(&settings, color);
+        let output = match paginate {
+            PaginationChoice::Always => UiOutput::new_paged_else_terminal(&settings),
```
Comment (Ralith) [inline src/ui.rs:147]: I don't think we should direct output to a pager unconditionally on startup. Long-running operations might want to display progress feedback directly to the terminal before pager-appropriate data is available, for example. I instead interpret "paginate always" the same way as "use color always", i.e. meaning "regardless of whether stdout looks like a tty."

File: src/ui.rs (hunk 14 above)
Comment (martinvonz) [reply src/ui.rs:172]: We won't spawn a pager anyway if `self.output` is a `UiOutput::Paged`. I prefer comparing to `Never` because it's more future-proof, but an assertion would also work (safer than a comment)

File: src/ui.rs (hunk 13 above)
Comment (martinvonz) [reply src/ui.rs:178]: To clarify, the reason I suggested `atty::is(stdout)` was that if `stdout` was not actually connected to stdout, then it would probably be better to check whatever it was connected to. But since that's not how `atty::is()` works, that's clearly not an option :)

I would still prefer to restructure the conditions using `match` instead of `matches!()`

File: src/ui.rs (hunk 17 above)
Comment (Ralith) [reply src/ui.rs:147]: Which is to say, we should still defer spawning the pager until requested.

Comment (martinvonz) [reply src/ui.rs:147]: Heh, I thought the same thing (I think I may have told @chooglen in person). But does it make sense to page when stdout doesn't look like a tty? Are there cases where that's useful?

Comment (chooglen) [reply src/ui.rs:147]: > I instead interpret "paginate always" the same way as "use color always", i.e. meaning "regardless of whether stdout looks like a tty."

Ah okay. I think the main issue (and the one I'm trying to address) is "what is the set of commands that trigger pagination?", which makes me think that I should just rename the config option to match that.

File: src/ui.rs (hunk 13 above)
Comment (Ralith) [reply src/ui.rs:178]: I don't think the `Stdout` type can represent anything but the hardcoded file descriptor, anyway.

File: src/ui.rs (hunk 17 above)
Comment (martinvonz) [reply src/ui.rs:147]: Oh, and I also checked that `hg` does the same thing. So does `git`:
```
$ PAGER='echo foo' git -p rm foo
foo
```
It doesn't even mention the error (there's no `foo` file in my repo)

File: src/ui.rs (hunk 14 above)
Comment (chooglen) [reply src/ui.rs:172]: Yeah I think comparing to `Never` is the most obvious approach.

File: src/ui.rs (hunk 17 above)
Comment (chooglen) [reply src/ui.rs:147]: > Ah okay. I think the main issue (and the one I'm trying to address) is "what is the set of commands that trigger pagination?", which makes me think that I should just rename the config option to match that.

Actually, this is probably better done via a CLI flag. I'll send that in a follow-up PR.

Comment (martinvonz) [reply src/ui.rs:147]: What would that flag be? Whatever it is, I agree to leave it for later so we the base feature in.

Comment (chooglen) [reply src/ui.rs:147]: In the follow up PR, I'm thinking of adding `--[no-]pager`, which is a bit like [hg](https://www.mercurial-scm.org/doc/hg.1.html). `hg` also accepts `--pager=(always|never|auto)` (which is what the config option is based on), but I'll drop it since it's confusingly similar-sounding to `ui.color`, but uses completely different rules.

Comment (Ralith) [reply src/ui.rs:147]: > what is the set of commands that trigger pagination?

I'd be surprised for pagination to ever be applied to code that doesn't have specific hardcoded support for it. Paginating terse status feedback would be strange.

> But does it make sense to page when stdout doesn't look like a tty?

I speculate that there's cases where stdout isn't technically a tty but something pager-like might still be useful. Perhaps a graphical pager when `jj` is invoked as a subprocess? Or perhaps tty cases where stdout is being mangled in some way that the user's configured pager is smart enough to bypass. I recall we saw `less` has a bunch of tricks to find the "real" tty.

Comment (chooglen) [reply src/ui.rs]: Okay, in the latest version:

- The pager is not always created on startup, only certain commands can spawn it.
- Added a CLI option that lets users disable the pager.
- `ui.paginate` is gone. We can figure out what to do about `always|never|auto` in a later PR.

File: src/ui.rs (hunk 13 above)
Comment (chooglen) [reply src/ui.rs]: The match was done. `atty` got replaced by `crossterm`

**Comment by chooglen** [comment]
In a future PR, I want to be able to paginate `jj --help`. It's easy to paginate if I add `request_pager()` call where we handle `clap` errors, but the tricky part is getting `jj --help --no-pager` to work right, since `--help` short-circuits the arg parsing.

File: tests/test_global_opts.rs [hunk 18]
```diff
@@ -253,6 +253,7 @@ fn test_help() {
           --no-commit-working-copy       Don't commit the working copy
           --at-operation <AT_OPERATION>  Operation to load the repo at [default: @] [aliases: at-op]
           --color <WHEN>                 When to colorize output (always, never, auto)
+          --no-pager                     Disables the pager
```
Comment (martinvonz) [inline tests/test_global_opts.rs:256]: nit: we seem to use imperative mood so s/Disables/Disable/

**Review by martinvonz — APPROVED** [review]
Thanks!
